libs/community/langchain_community/retrievers/product_quantization.py

```diff
@@ -0,0 +1,118 @@
+from __future__ import annotations
+
+import concurrent.futures
+from typing import Any, Iterable, List, Optional
+
+import numpy as np
+from langchain_core.callbacks import CallbackManagerForRetrieverRun
+from langchain_core.documents import Document
+from langchain_core.embeddings import Embeddings
+from langchain_core.retrievers import BaseRetriever
+
+
+def create_index(contexts: List[str], embeddings: Embeddings) -> np.ndarray:
+    """
+    Create an index of embeddings for a list of contexts.
+
+    Args:
+        contexts: List of contexts to embed.
+        embeddings: Embeddings model to use.
+
+    Returns:
+        Index of embeddings.
+    """
+    with concurrent.futures.ThreadPoolExecutor() as executor:
+        return np.array(list(executor.map(embeddings.embed_query, contexts)))
+
+
+class PQRetriever(BaseRetriever):

```

Comment:
could we call this something like NanoPQRetriever to make clear what dependency is being used?

Reply:
Felt like the algorithm's name should be given more importance than the package. Thought similar on the lines of TF-IDF, SVM or KNN

Reply:
think since this package is less well known/used than sklearn its more important to make it obvious whats being used

Reply:
Understood, will update on this.